[BE] Project Review 요청

haveagood opened the pull request:
- 현재까지의 진행상황
1. Main Page API 개발 완료
2. Filter 기능 (Main Page 내부) 구현 완료
3. Detail Page 기능 구현 완료
4. DB 스키마 설계 완료
5. Sample Data 완료

BackEnd Review 요청
Members : Alex, Hamill

haveagood committed:
[chore] dev branch 생성 및 be 프로젝트 폴더 생성

haveagood committed:
[BE] feat : CORS에러 방지를 위한 filter와 webConfig 추가

- CORS 에러 방지를 위해서 CORS filter, WebConfig 추가

haveagood committed:
[BE] feat : DB 설계 및 Schema 생성

- DB 구조 설계
- schema 파일 생성

issue : #2

-------------------------
-------------------------

hanurii committed:
Merge pull request #11 from codesquad-member-2020/feature/be/schema

[BE] feat : DB 설계 및 Schema 생성

hanurii committed:
[BE] fix : Project SDK 변경

- jdk 1.8.0_242 -> jdk 1.8.0_231 로 변경

haveagood committed:
[BE] feat : schema 등록 및 EC2 서버에서 mysql 등록 확인

- shcema 등록
- RDS 터널링 실패로 인해 EC2 서버에 Mysql DB 생성
- DataSource로 connection확인

haveagood committed:
[BE] chore : schema 수정

- iOS와 Mock API에 대해서 논의 후, 데이터의 변수명을 변경하였음
- 해당 변경사항을 바탕으로 schema 변수 수정

-------------------------
-------------------------

haveagood committed:
Merge branch 'dev-be' into feature/be/schema

haveagood committed:
[BE] chore : conflict 해결 및 스키마 수정

- 충돌 해결 및 스키마 수정

-------------------------
-------------------------

hanurii committed:
[BE] feat : workspace.xml 추가

haveagood committed:
[BE] chore : hotel data csv 파일 추가

- 외부에서 수집한 airbnb 데이터를 resources/data 디렉토리에 추가
- csv 파일 형태로 mysql에 추가할 예정

-------------------------
-------------------------

haveagood committed:
[BE] chore : csv 파일 수정

- csv 파일 중복 및 null 제거 및 수정

-------------------------
-------------------------

haveagood committed:
[BE] chore : schema , data 수정 및 DATA INSERT query문 작성

- schema 형식을 data 형식에 맞춰서 수정하였습니다.
- init data 수정
- data를 넣기위한 쿼리문 작성

# 본문 작성

# 꼬릿말은 아래에 작성 : ex) #이슈 번호

# --COMMIT END--
# <타입> 리스트
# feat : 기능 (새로운 기능)
# fix : 버그 (버그 수정)
# refactor : 리팩토링
# style : 스타일(코드 형식, 세미콜론 추가 : 비지니스 로직 변경 X)
# docs : 문서 (문서 추가, 수정, 삭제)
# test : 테스트 (테스트 코드 추가, 수정, 삭제 : 비지니스 로직 변경 X)
# chroe : 기타 변경사항(빌드 스크립트 수정 등)
-------------------------
#	제목 첫 글자 대문자
#	제목은 명령문
#	제목 끝에 마침표(.) 금지
#	제목과 본문을 한 줄 띄워 분리하기
#	본문은 "어떻게" 보다 "무엇을" , "왜"를 설명한다
#	본문에 여러줄의 메시지를 작성할 땐 "-"로 구분
-------------------------

haveagood committed:
Merge pull request #21 from codesquad-member-2020/feature/be/schema

- DB 설계
- DB schema 작성
- Data 수집
- Data Insert

haveagood committed:
[BE] chore : workspace 파일 삭제 및 gitignore 추가

- 불필요하게 workspace파일이 계속 변경되고, git에 올라가서 삭제하였습니다.
- 더이상 올라가지 않도록 gitignore에 추가하였습니다.

# 본문 작성

# 꼬릿말은 아래에 작성 : ex) #이슈 번호

# --COMMIT END--
# <타입> 리스트
# feat : 기능 (새로운 기능)
# fix : 버그 (버그 수정)
# refactor : 리팩토링
# style : 스타일(코드 형식, 세미콜론 추가 : 비지니스 로직 변경 X)
# docs : 문서 (문서 추가, 수정, 삭제)
# test : 테스트 (테스트 코드 추가, 수정, 삭제 : 비지니스 로직 변경 X)
# chroe : 기타 변경사항(빌드 스크립트 수정 등)
-------------------------
#	제목 첫 글자 대문자
#	제목은 명령문
#	제목 끝에 마침표(.) 금지
#	제목과 본문을 한 줄 띄워 분리하기
#	본문은 "어떻게" 보다 "무엇을" , "왜"를 설명한다
#	본문에 여러줄의 메시지를 작성할 땐 "-"로 구분
-------------------------

haveagood committed:
[BE] feat : PropertiesDao 생성

- PropertiesDao 클래스 생성

issue : #24

# 본문 작성

# 꼬릿말은 아래에 작성 : ex) #이슈 번호

# --COMMIT END--
# <타입> 리스트
# feat : 기능 (새로운 기능)
# fix : 버그 (버그 수정)
# refactor : 리팩토링
# style : 스타일(코드 형식, 세미콜론 추가 : 비지니스 로직 변경 X)
# docs : 문서 (문서 추가, 수정, 삭제)
# test : 테스트 (테스트 코드 추가, 수정, 삭제 : 비지니스 로직 변경 X)
# chroe : 기타 변경사항(빌드 스크립트 수정 등)
-------------------------
#	제목 첫 글자 대문자
#	제목은 명령문
#	제목 끝에 마침표(.) 금지
#	제목과 본문을 한 줄 띄워 분리하기
#	본문은 "어떻게" 보다 "무엇을" , "왜"를 설명한다
#	본문에 여러줄의 메시지를 작성할 땐 "-"로 구분
-------------------------

haveagood committed:
[BE] chore : git ignore 오타 수정

- git ignore 오타 수정

# 본문 작성

# 꼬릿말은 아래에 작성 : ex) #이슈 번호

# --COMMIT END--
# <타입> 리스트
# feat : 기능 (새로운 기능)
# fix : 버그 (버그 수정)
# refactor : 리팩토링
# style : 스타일(코드 형식, 세미콜론 추가 : 비지니스 로직 변경 X)
# docs : 문서 (문서 추가, 수정, 삭제)
# test : 테스트 (테스트 코드 추가, 수정, 삭제 : 비지니스 로직 변경 X)
# chroe : 기타 변경사항(빌드 스크립트 수정 등)
-------------------------
#	제목 첫 글자 대문자
#	제목은 명령문
#	제목 끝에 마침표(.) 금지
#	제목과 본문을 한 줄 띄워 분리하기
#	본문은 "어떻게" 보다 "무엇을" , "왜"를 설명한다
#	본문에 여러줄의 메시지를 작성할 땐 "-"로 구분
-------------------------

haveagood committed:
Merge pull request #25 from codesquad-member-2020/feat/be/dao

Feat/be/dao

haveagood committed:
[BE] chore : 인텔리제이 실행 시, DB 초기화 되지 않도록 코드 수정

- 스프링부트 실행시 DB 초기화를 시키지 않도록 schema.sql을 data source에서 읽지 않도록 수정

hanurii committed:
[BE] chore : properties.csv 데이터 수정

haveagood committed:
[BE] feat : Controller 생성

- 초기 페이지 화면을 호출하는 API를 구현하기 위해 Controller 생성
- 인자값을 Path로 받았습니다.

issue : #5

hanurii committed:
Merge pull request #28 from codesquad-member-2020/feat/be/main-api-hamill

[BE] chore : properties.csv 데이터 수정

haveagood committed:
[BE] feat : Controller에서 호출되는 Service 생성

- Controller에서 Dao로 바로 접근하는 것이 아니라, service를 통하도록 구현하였습니다.
- Path값으로 받은 PageNumber에 20을 곱하여 propertyRange를 설정하였습니다.
- 1을 요청하면 id값이 1~19인 데이터를, 2를 요청하면 1~39... 순차적으로 값이 증가하도록 구현하였습니다.

issue : #5

haveagood committed:
[BE] feat : 값을 저장하는 Dto 생성

- Main Page 구현에 필요한 인자값을 담는 Dto 클래스를 생성하였습니다.

- Lombok을 활용해서 Builder를 사용하였습니다.

issue : #5

haveagood committed:
[BE] feat : Dao 구현

- Main페이지 요청이 들어왔을 때, DB에서 값을 받아오는 Dao클래스를 구현하였습니다.
- group_concat을 활용하여 one-to-many값을 받아와서 List로 저장하였습니다.

issue : #5

haveagood committed:
[BE] chore : class 분리

- Pair프로그래밍을 하다보니, 각자 프로젝트를 개발하자고 의견 통일
- Alex,Hamil클래스를 나눠서 관리해보기로 결정!!

-------------------------
-------------------------

haveagood committed:
[BE] chore : csv 데이터 수정 및 data.sql 수정

- springboot 실행 시, LOAD DATA 자동 실행되게 하기 위한 수정

-------------------------
-------------------------

haveagood committed:
[BE] feat : Login기능을 위한 GitOAuth 연동

- Oauth 관련 데이터를 주고받을 Dto 생성
- LoginService 구현
- Git Oauth ClientID, ClientSecret application-secret에 별도로
  저장(git에는 노출되지 않도록)

-------------------------
-------------------------

haveagood committed:
[BE] feat : github Oauth 로그인 기능 구현

- 현재, CORS에러가 나는지 확인을 못한 상태
- USER의 정보가 잘 전달되는지도 확인 못한 상태
- local test결과 user id를 쿠키에 저장하는것 까지 확인
- githubUser 객체에 userid, githubid, email을 가지고 있게 하였음
- 해당 데이터를 dao를 통해서 user라는 table에 저장할 예정

-------------------------
-------------------------

haveagood committed:
[BE] feat : oauth 데이터 db에 저장하는 기능 구현

- github user가 oauth를 통해서 들어온 경우 해당 user의 데이터를 db에
  저장하도록 구현

- 만약 존재하는 경우 중복을 방지하기 위해서 저장하지 않도록 구현

-------------------------
-------------------------

haveagood committed:
[BE] feat : TestCode 작성

- properties의 결과물에 대한 테스트 코드 작성
- user데이터에 대한 테스트 코드 작성

-------------------------
-------------------------

hanurii committed:
[BE] modify : mysql 로 접속하는 ip 주소 변경

hanurii committed:
[BE] feat : AirbnbController 클래스 생성

hanurii committed:
[BE] modify : class 이름 변경

hanurii committed:
[BE] feat : AirbnbServiceHamill 클래스 추가

haveagood committed:
[BE] feat : DB Detail구조 변경 및 DATA 추가

- detail 구조 변경 및 Data 추가

-------------------------
-------------------------

haveagood committed:
[BE] chore : Detail Table 스키마 변경 및 데이터 삽입

- Detail 테이블 스키마 변경 및 데이터 구조 통일

-------------------------
-------------------------

haveagood committed:
[BE] chore : schema 수정

- schema 타입 수정

-------------------------
-------------------------

hanurii committed:
[BE] feat : csv 파일 자동 insert 구현

- csv 파일을 웹 서버 안 db 루트에 넣고 intellij 구동 시 csv 파일이 db에 들어갈 수 있게 구현
- csv 파일은 총 2개로 properties.csv와 properties_detail.csv 파일이다

hanurii committed:
[BE] feat : images.csv 파일 웹서버 db insert 구현

hanurii committed:
[BE] feat : Properties DTO 생성 및 클래스명 변경

- issue : #5

hanurii committed:
[BE] feat : 홈 화면 api 구현

- alex 가 구현한 코드를 참고하여 rest api 구현

- issue : #5

hanurii committed:
[BE] feat : 홈 화면 밑으로 스크롤 시 추가 데이터 전송

- url 파라미터로 page 값을 주면 그 값 * 20 에 해당하는 방에 대한 정보를 보내준다

- issue : #5

hanurii committed:
[BE] modify : properties 테이블의 address 컬럼 삭제

- 불필요한 컬럼을 삭제함

hanurii committed:
[BE] feat : detail 페이지 dao 구현

- issue : #7

hanurii committed:
[BE] feat : detail 페이지 service, controller 구현

- issue : #7

hanurii committed:
[BE] modify : price 와 관련된 변수 이동

- issue : #7

hanurii committed:
[BE] feat : 요금 필터 기능 구현

- 최소 금액과 최대 금액을 정해줄 시 그에 맞는 숙소들을 보여준다

- issue : #6

hanurii committed:
[BE] feat : 인원 필터 기능 구현

- adults와 children 인원 수를 합친 값을 기준으로 숙소 목록을 출력한다

- issue : #6

haveagood committed:
[BE] refactor : main page filter 기능 구현

- 기존의 main 화면에서 RequestParameter를 받아서 해당 조건에 맞춰서 값을
  필터링하는 기능 구현
- 우선적으로 숙박 인원수를 필터링하여 가지게하는 코드 구현
- 변수를 유동적으로 받아서 처리하는 방법 고민중

issue : #5,#6

-------------------------
-------------------------

haveagood committed:
[BE] feat : checkin, checkout 별로 필터링 기능 구현

- 날짜가 입력되지 않으면 현재의 날짜를 표시하도록 구현
- calender 테이블에 booking이 되지 않는 숙소만 나오도록 구현

issue : #5, #6

-------------------------
-------------------------

haveagood committed:
[BE] feat : min,max Price에 따른 데이터 필터링 기능 구현

- min, max Price에 따른 필터링 기능 구현
- 우선적으로 Controller에서 문자열로 값을 입력받도록 하였다.
- 혹시모를 RequestParam조작을 대비하여 문자열로 받고, Controller에서
  파싱하도록 하였다.
- 파싱하는 과정에서 에러가 발생할 경우 min, max값을 DEFAULT로 설정한
  값으로 가도록 하였다.

issue : #5, #6

-------------------------
-------------------------

hanurii committed:
[BE] feat : calendar 테이블 추가

- issue : #6

hanurii committed:
[BE] feat : 요금/인원 exception 처리하는 메서드 구현

- url 파라미터가 잘못된 값으로 들어오더라도 디폴트 값을 넣어서 서비스에 지장없도록 구현

- issue : #6

haveagood committed:
[BE] feat : Detail Page 기능 구현

- Detail Page의 경우에는 표시해야 할 자료가 너무 많았다.
- 1차적으로 상속을 활용하여 구현하고자 하였는데, table단위로 객체를
  나눠서 상속을 하고자 하였지만, 그럼에도 변수가 너무 많았고, 가독성이
  떨어졌다.
- Component를 활용하여 host,price,location,review,room 으로 객체를
  분리하고 해당되는 변수들을 해당 객체에 이동시켰다.
- DetailDto라는 Main객체에서 하위 컴포넌트를 포함하게 하여 객체를
  생성하였다.

issue : #7

-------------------------
-------------------------

haveagood committed:
[BE] refactor : host_type 변환

- DB에 존재하는 host_type의 형식이 String형식으로 ex)'super'로 되어있다
- 확장성을 고려하여 문자열로 받고자 하였으나, 프로젝트에서는 슈퍼호스트
  여부만 판단하면 될 것 같다고 결정하였다.
- DB의 구조를 바꾸기보다는 CASE WHEN sql구문을 활용하여 값을 변환시켰다.
- 기존의 값은 string이었는데, 1,0값을 가지는 boolean으로 변환하였다.

issue : #7

-------------------------
-------------------------

haveagood committed:
[BE] refactor : amenities column 쿼리 수정

- amenities 데이터가 ""로 이루어져있어 DB에서 가져올때, 깔끔하게
  가져와지지 않았다.
- REPLACE를 활용하여 Data를 정리해서 가져오도록 변환

issue : #7

-------------------------
-------------------------

hanurii committed:
[BE] feat : BETWEEN 함수 사용과 price_min 보다 price_max가 작을 때 생기는 예외 처리

- issue : #6

haveagood committed:
[BE] feat : 좌표정보 전송 API 개발

- Controller에 API 추가

-------------------------
-------------------------

hanurii committed:
[BE] feat : 날짜 필터 기능 구현

- 날짜가 넘어오면 예약이 가능한 숙소만 보여지게 구현
- checkin 보다 checkout이 과거 날짜로 넘어와도 제대로 checkin날짜와 checkout 날짜가 전후 관계가 될 수 있도록 구현

- issue : #6

haveagood committed:
[BE] refactor : main 페이지 filter 기능 수정

- main 페이지에서 날짜값이 큰 값이 먼저 오는 경우 비교를 통해서 작은
  날짜가 우선적으로 오도록 메소드 구현

-------------------------
-------------------------

hanurii committed:
Merge pull request #58 from codesquad-member-2020/feat/be/rest-api-practice

[BE] Hamill이 개발한 브랜치를 merge합니다

haveagood committed:
Merge branch 'feat/be/geo,map-alex' into dev-be

haveagood committed:
[BE] refactor : complict 해결

- code merge 단계에서 발생하는 conflict 해결

-------------------------
-------------------------

hanurii committed:
[BE] modify : hamill 환경에 맞게 속성값과 data.sql 변경

hanurii committed:
[BE] feat : detail 페이지 JSON 응답 객체 형태를 변경하기 위해 Composition 생성 후 detail 페이지와 관련된 소스들 변경

- PropertiesDaoHamill : findByPropertiesId 내용 변경
  - alex 가 구현한 코드를 참고하여 구현

- issue : #7

wheejuni commented on be/.idea/compiler.xml:
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
`.idea` 하위의 파일들이 git에 들어오는 것과 안 들어오는 것의 장단점이 있기는 한데요.
저는 빼는 걸 선호하긴 해요. 두 경우 차이에 대해 학습해보시고, 스스로 판단해 보셨음 좋겠네요. 
지금 커밋된 파일의 분량이 많기는 해요.

wheejuni commented on be/src/main/java/com/airbnb3/codesquad/airbnb3/dao/DetailDaoAlex.java:
```diff
@@ -0,0 +1,84 @@
+    private static final Logger logger = LoggerFactory.getLogger(DetailDaoAlex.class);
+
+    @Autowired
+    NamedParameterJdbcTemplate jdbcTemplate;
```
**접근 제어자** 잊지 말아주시고요.

wheejuni commented on be/src/main/java/com/airbnb3/codesquad/airbnb3/dao/DetailDaoAlex.java:
```diff
@@ -0,0 +1,84 @@
+                                .numberOfReviews(rs.getInt("number_of_reviews"))
+                                .reviewScoresAccuracy(rs.getDouble("review_scores_accuracy"))
+                                .reviewScoresCheckin(rs.getDouble("review_scores_checkin"))
+                                .reviewScoresCleanliness(rs.getDouble("review_scores_cleanliness"))
```
저도 람다식을 선호하긴 하는데요,
이정도 복잡도라면 별도 클래스로의 분리를 고려해봄직 한 것 같아요.

wheejuni commented on be/src/main/java/com/airbnb3/codesquad/airbnb3/dao/PropertiesDaoHamill.java:
```diff
@@ -0,0 +1,143 @@
+        logger.info("#### priceMax: {}", priceMax);
+        logger.info("#### checkIn: {}", checkIn);
+        logger.info("#### checkOut: {}", checkOut);
+        logger.info("#### accommodates: {}", accommodates);
```
이런 목적으로 `INFO` 로그는 너무 길고 빈번해요. 
한 번의 `logger.info` 로 다 로깅하거나 아예 로깅하지 않는 것을 추천해요. 
`logger.debug` 정도도 괜찮겠네요. 로컬에선 root level을 debug로 주고, EC2 상에서는 INFO 이상으로.

wheejuni commented on be/src/main/java/com/airbnb3/codesquad/airbnb3/dao/PropertiesDaoHamill.java:
```diff
@@ -0,0 +1,143 @@
+
+    private List<String> imageParser(String images) {
+        String[] image = images.split(",");
+        return Arrays.asList(image);
```
이렇게 리팩토링해보면 어떄요?

```suggestion
        return Arrays.asList(images.split(",");

wheejuni commented on be/src/main/java/com/airbnb3/codesquad/airbnb3/dao/UserDaoAlex.java:
```diff
@@ -0,0 +1,25 @@
+public class UserDaoAlex {
+
+    @Autowired
+    JdbcTemplate jdbcTemplate;
```
**접근 제어자** 잊지맙시다. 
자꾸 잊는다면 https://namu.wiki/w/%EC%A7%80%EC%A0%81%ED%99%95%EC%9D%B8%20%ED%99%98%ED%98%B8%EC%9D%91%EB%8B%B5 이걸 해보시는 것도 좋습니다. 코드를 손으로 가리키면서 "접근 제어자!" 라고 외치는거죠. 

농담이구요. 이제는 정말 잊지 맙시다.

haveagood replied:
정말 ㅠㅠ 잊지않겠습니다 감사합니다 ㅋㅋ

wheejuni commented on be/src/main/java/com/airbnb3/codesquad/airbnb3/dto/DetailDtoAlex.java:
```diff
@@ -0,0 +1,38 @@
+    private Boolean saved;
+    private List<String> images;
+    // Composition (비슷한 변수들 별도의 객체로 분류)
+    private HostDto hostInfo;
+    private LocationDto locationInfo;
+    private PriceDto priceInfo;
+    private ReviewDto reviewInfo;
+    private RoomDto roomInfo;
```
이런 구성 좋습니다.

haveagood replied:
추가적인 질문이 있습니다!

전달해야하는 데이터의 양이 너무 방대하여 컴포지션을 활용하여 데이터를 나눴습니다.

1. 제가 사용한 방식이 컴포지션이 맞다고 할 수 있는지가 궁금합니다.

2 .추가적으로, 상속과 컴포지션의 관계를 공부하다보니 상속보다는 컴포지션을 활용하는 이유가,

 상위클래스가 변경이되면 그것이 하위 클래스에 영향을 주고, 캡슐화가 깨지는 이유인것 같습니다.

만약, Dto로 단순하게 데이터를 담는 객체의 경우 객체가 변경되는 일이 흔하지는 않은데, 이런 경우에는 상속을 활용해도 좋을까요?

(이것을 고민하는 이유는, DetailDto라는 객체가 가지고 있는 정보가 PropertiesDto라는 객체가 가지고 있는 정보를 포함하여 가지고 있기 때문입니다!)

wheejuni replied:
> 만약, Dto로 단순하게 데이터를 담는 객체의 경우 객체가 변경되는 일이 흔하지는 않은데, 이런 경우에는 상속을 활용해도 좋을까요?

**오해죠. 엄청나게 흔합니다.** 
단순한 API상의 응답일수록 변경이 더 쉽고 그만큼 변경도 더 빈번할 겁니다. 

그래서 지금 구현하신 방향이 틀리지 않다고 생각해요. 
하위 필드 집합만 바뀐다면, 그 DTO 클래스만 바꾸면 되니까요.

wheejuni commented on be/src/main/java/com/airbnb3/codesquad/airbnb3/dto/PropertiesDetailDtoHamill.java:
```diff
@@ -0,0 +1,52 @@
+@Setter
+@ToString
+@Builder
+public class PropertiesDetailDtoHamill {
```
정말 방대하네요. 정말 한번에 전달해야 할 정보인가요? 모두 다?
다시 한 번 고민해보시길 추천 드립니다.

wheejuni commented on be/src/main/java/com/airbnb3/codesquad/airbnb3/dto/composition/HostDto.java:
```diff
@@ -0,0 +1,26 @@
+    private String hostAbout;
+
+    @Builder
+    public HostDto(Boolean isSuperHost, String notes, String hostName, String hostSince, String hostLocatison, String hostAbout) {
```
파라미터 수가 적지 않은데, 생성자를 `private` 으로 가리는 건 어떤가요? 
빌더만 사용되게요.

haveagood replied:
파라미터의 수에 따라 생성자를 private 혹은 public으로 다르게 관리하는 이유가 있나요?
아니면 일단 빌더를 사용하면 생성자는 private으로 숨기는것이 좋을까요?

wheejuni replied:
@haveagood 파라미터의 숫자가 많아질수록 데이터타입이 겹칠 확률도 높고, 이렇게 되면 휴먼 에러가 날 가능성도 그만큼 계속 높아집니다.

예를 들어 "연애사" 라는 클래스를 만들었다고 생각해보죠. 
제 전 여친이 "나연" 이고 현 여친이 "슬기" 라고 가정해 보겠습니다. 

```java
public class RomanceHistory {
    private String exGirlFriend;
    private String currentGirlFriend;

    public RomanceHistory(String exGirlFriend, String currentGirlFriend) {
        this.exGirlFriend = exGirlFriend;
        this.currentGirlFriend = currentGirlFriend;
    }

...
}
```

이런 클래스가 나오겠죠. 그리고 아래와 같은 실수를 했다고 가정해 봅시다. 

`RomanceHistory history = new RomanceHistory("슬기", "나연");`

현 여친 자리에 전 여친을 넣었고 전 여친 자리에 현 여친을 넣었는데 **아무런 컴파일 에러도 나지 않고 시스템적으로 검증할 방법이 없습니다.** 둘 다 `String` 이니까요. 

아래와 같은 빌더가 있으면 안전하겠죠. 

```java
RomanceHistory history = RomanceHistory.builder().currentGirlFriend("슬기").exGirlFriend("나연").build();
```

생성자의 파라메터가 많으면 많을 수록 이런 실수를 할 가능성이 많기 때문에 `builder` 를 잘 설계하고 가급적 생성자를 `private` 하게 가리는 구현을 권하는 것입니다.

wheejuni commented on be/src/main/java/com/airbnb3/codesquad/airbnb3/service/AirBnbService.java:
```diff
@@ -0,0 +1,102 @@
+        Date checkInDate = parseStringToCheckInDate(checkIn);
+        Date checkOutDate = parseStringToCheckOutDate(checkOut);
+
+        if (checkInDate.compareTo(checkOutDate) > 0) {
```
음... 이건 어떤 구현인가요?
이것보다는 https://docs.oracle.com/javase/8/docs/api/java/util/Date.html 을 참고해서...
`.before(Date date)` , `.after(Date date)` 라는 메소드가 있네요. 이걸 활용하셔야 할 것 같아요.

haveagood replied:
이번 프로젝트가 숙소 예약 프로젝트입니다.
그래서, 날짜를 입력받을 때, 혹시모를 URL 조작에 따른 에러를 방지하기 위해서
Controller에서 문자열로 날짜 데이터를 입력받고, 문자열을 Date Type으로 치환을 하는것에 대한 구현인 것 같습니다.
추가적으로 저희가 DB에서 데이터 조회를 BETWEEN을 활용하여 조회를 하는데, BETWEEN A AND B 라는 문법에서 A의 값이 B보다 크면 에러가 발생하기 때문에, 
만약 데이터가 거꾸로 들어왔을 때, A.compareTo(B)라는 문법으로 Date Type의 대소비교를 하는 방법을 찾았습니다.
그래서 두개의 날짜를 비교하고, A의 값이 클 경우에 순서를 바꿔주는 구현입니다.
compare 대신에 before와 after를 공부하여 수정하겠습니다. 감사합니다!

haveagood replied:
현재 숙소 예약에서 시간을 고려하지 않고, 날짜만을 고려하고있어서 Date 타입(java.sql.Date)를 사용하고 있습니다. (sql쿼리문에 연동하기 위해서, LocalDate는 쿼리에 바로 넣는것이 안되는걸로 알고있음)

기존에는 Date 대신 LocalDate를 계속해서 활용해왔는데, 계속 LocalDate를 활용하는것이 좋을까요?
LocalDate를 활용한다면, LocalDate를 format을 통해서 2020-05-01 이런 문자열로 매칭을 해보려고 합니다.

wheejuni reviewed (commented):
코드 작성하시느라 수고 많으셨어요. 👍 
프로젝트가 복잡하고 코딩량이 방대하다보니 한번 훑어보는 것도 쉽지만은 않네요. 

일단 눈에 보이는 수정 포인트만 잡아서 코멘트 드렸습니다.
확인해 주시고, 다른 질문 있으시면 다시 알려주시고요. 감사합니다!

hanurii committed:
[BE] modify : IPv4 퍼블릭 IP가 변경됨에 따라 application.properties에 들어있던 ip 주소 값도 변경

hanurii committed:
[BE] feat : EC2-RDS 서버 배포 성공

- IPv4 : 52.78.52.252
- RDS : airbnb-rds.ct12sp1kygt2.ap-northeast-2.rds.amazonaws.com
- CSV 파일은 수동으로 넣어주어야 한다
  - 수동으로 매번 넣기 힘들기 때문에 한 번 넣어주고 그 뒤로 테이블을 DROP 하지 않는다

- issue : #74

hanurii committed:
[BE] modify : host_type(String) 을 is_super_host (Boolean) 로 변경

- issue : #5

hanurii committed:
[BE] feat : offset * 20 으로 변경

- issue : #5

hanurii committed:
[BE] feat : null 체크를 원활히 하기 위해 @RequestParam 속성값인 defaultValue 추가

haveagood committed:
[BE] feat : 접근제어자 설정 및 좌표값 받아오는 코드 작성

- 코드리뷰에서 받은 접근제어자를 설정하였습니다.
- 좌표값의 범위를 받아서 해당 범위의 숙소를 조회하는 기능을
  추가하였습니다.

-------------------------
-------------------------

haveagood committed:
[BE] chore : images.csv 수정 및 api 네이밍 변경

- images.csv에 조금 더 많은 데이터 추가
- api 네이밍 변경

-------------------------
-------------------------

haveagood committed:
[BE] chore : images 파일 수정

- images 파일 수정

-------------------------
-------------------------

haveagood committed:
[BE] feat : 예약하는 기능 구현 시작

- reservation service 분리
- reservation dto, dao 분리
- 일차적으로 cookie에서 user 정보값을 가져온다고 생각하고 controller에서 @cookievalue를 활용, default값을 none으로 하였습니다.
- reservation service에서 properties_id, checkin, checkout,name을 받아서
  해당 properties에 대한 예약을 진행하고자 합니다.

issue : #10

-------------------------
-------------------------

haveagood committed:
[BE] feat : reservationDto 생성

- reservation에 필요한 DTO를 만들고, 그에 필요한 데이터를 Service에서
  생성하여 Dto에 builder로 만들었습니다.

- reservationDao에 해당 객체를 집어넣어서 Insert를 할 생각입니다.

issue : #10

-------------------------
-------------------------

haveagood commented:
추가적으로 질문 하나 남기고 싶습니다!!

parameterSource.addValues(new ObjectMapper().convertValue(reservationDto,Map.class));
이런 식으로 객체를 ObjectMapper 를 활용 하여 Map으로 변환한 뒤에
한번에 sql 의 parameter 로 변환을 할 수 있는 것 같습니다.

하지만, 가독성 측면에서 어떤 파라미터가 존재하는지 알 수 없는 상황이 되어버리는데
이런 경우에는 Dto에서 getter를 활용하여 각 변수들을 sql의 parameter에 직접 할당하여 
어떤 변수들이 파라미터로 존재하는지 코드상에서 표현하는 것이 바람직한가요?

NamedParameterJdbcTemplate을 사용하고 있어서 sql에서 parameter의 변수이름을 명시하긴 합니다.

- 개인적인 생각은 parameter의 값을 명시하는것이 가독성 측면에서 좋다고 생각하는데 sql에서 어떤 parameter를 사용하는지 표시가 되는데 안써도 되는건가? 라는 의문이 들어서 질문 남깁니다.

예를 들면, 이런식으로 parameter를 할당하는 방식과
<img width="570" alt="스크린샷 2020-05-28 오후 9 36 10" src="https://user-images.githubusercontent.com/42365037/83141979-49ad9600-a12b-11ea-8239-620e2511c032.png">

이런식으로 객체에 넣어서 한번에 할당하는 방식
<img width="627" alt="스크린샷 2020-05-28 오후 9 37 28" src="https://user-images.githubusercontent.com/42365037/83142466-7feb1580-a12b-11ea-9e5a-f862317387c5.png">

하지만 사용하는 parameter를 sql문에서 이런식으로 명시하여 사용함
<img width="532" alt="스크린샷 2020-05-28 오후 9 38 23" src="https://user-images.githubusercontent.com/42365037/83142525-92fde580-a12b-11ea-958a-a7c7227aa979.png">

hanurii committed:
[BE] modify : imageParser() 삭제하고 builder 에서 수정함

- Arrays.asList(rs.getString("image").split(","))

- issue : #7

hanurii committed:
[BE] feat : log level 설정 (info -> debug)

- 배포된 서버에는 info를
- 연습으로 사용하는 log level은 debug 를

hanurii committed:
[BE] feat : 필요한 API 추가, booking 테이블 수정

- GET url/properties : 모든 숙소 목록 반환
- GET url/properties/{propertiesId} : id 값에 해당하는 숙소 반환
- PUT url/reservations/{reservationsId} : id 값에 해당하는 숙소 저장
- DELETE url/reservations/{reservationsId} : id 값에 해당하는 숙소 삭제
- GET url/reservations : 모든 예약 목록 반환

- issue : #60

haveagood committed:
[BE] feat : reservation기능 구현

- 예약기능 구현
- 프론트에서 받는 변수에 대해서는 아직 정해지지 않아서 임의로 변수를
  설정하여 만들었음
- 프론트쪽과 논의 필요

issue : #10

-------------------------
-------------------------

hanurii committed:
[BE] feat : 예약 기능 구현

- 예약 버튼을 눌렀을 때 예약 데이터를 저장합니다

- issue : #60

hanurii committed:
[BE] feat : 예약 취소 기능 구현

- reservation id 값을 가지고 예약 취소를 할 수 있다

- issue : #60

hanurii committed:
[BE] feat : 예약 성공, 예약 취소 때 사용할 메세지 클래스 생성

- 상태코드와 메세지를 객체 형태로 전달한다

- issue : #60

hanurii committed:
[BE] feat : 모든 예약 출력 기능 구현

- 예약된 숙소들을 출력합니다

- issue : # 10

hanurii committed:
[BE] modify : 데이터소스 재 연결과 코드 정렬

hanurii committed:
Merge branch 'feat/be/api-develop-hamill' into feat/be/deploy-hamill

# Conflicts:
#	be/src/main/resources/schema.sql

충돌 해결

haveagood committed:
[BE] feat : reservation기능 구현 완료

- 현재, 네이밍 수정중
- reservation 기능 구현 완료

issue : #10

-------------------------
-------------------------

hanurii committed:
Merge remote-tracking branch 'origin/feat/be/reservation-alex' into feat/be/api-develop-hamill

# Conflicts:
#	be/out/production/resources/application.properties
#	be/out/production/resources/schema.sql
#	be/src/main/resources/application.properties
#	be/src/main/resources/schema.sql

hanurii committed:
[BE] modify : Hamill, Alex코드 병합

- 계속 코드 수정중

haveagood committed:
[BE] chore : alex,hamill 디렉토리 생성

- code merge를 위해서 각자의 작업 영역을 각자 디렉토리로 이동
- 공통된 deploy용 객체 생성

-------------------------
-------------------------

hanurii committed:
[BE] modify : 컴포지션 코드 merge를 위해 수정

haveagood committed:
Merge branch 'feat/be/api-merge' of https://github.com/codesquad-member-2020/airbnb-03 into feat/be/api-merge

hanurii committed:
[BE] feat : 코드 merge를 통한 공통된 DTO 생성

haveagood committed:
[BE] refactor : properties list 출력하는 api 메소드 관련 리팩토링

- 불필요한 예외를 잡아내는 처리 제거
- Double Type의 경우 계산과 같은 로직에서 오차가 발생할 수 있기 때문에, BigDecimal 타입으로 모두 변경
- 좌표, 가격 데이터를 BigDecimal로 변경
- 기존 @Autowired를 활용했던 인젝션 방식을, 생성자 주입으로 변경

# 본문 작성

# 꼬릿말은 아래에 작성 : ex) #이슈 번호

# --COMMIT END--
# <타입> 리스트
# feat : 기능 (새로운 기능)
# fix : 버그 (버그 수정)
# refactor : 리팩토링
# style : 스타일(코드 형식, 세미콜론 추가 : 비지니스 로직 변경 X)
# docs : 문서 (문서 추가, 수정, 삭제)
# test : 테스트 (테스트 코드 추가, 수정, 삭제 : 비지니스 로직 변경 X)
# chroe : 기타 변경사항(빌드 스크립트 수정 등)
-------------------------
#	제목 첫 글자 대문자
#	제목은 명령문
#	제목 끝에 마침표(.) 금지
#	제목과 본문을 한 줄 띄워 분리하기
#	본문은 "어떻게" 보다 "무엇을" , "왜"를 설명한다
#	본문에 여러줄의 메시지를 작성할 땐 "-"로 구분
-------------------------

haveagood committed:
[BE] refactor : properties list 메소드 수정 완료

- filter기능 테스트 아직
- 기본적으로 spring-boot가 돌아가는지 확인하기 위해서 에러가 발생하는
  객체 주석처리하여 러닝

-------------------------
-------------------------

haveagood committed:
[BE] refactor : properties list 조회 기능 수정

- 조회시 요청하는 user의 id에 따라서 saved표시가 다르게 나타나도록 수정
- 즐겨찾기가 각 사용자별로 다르게 표시되게 구현

-------------------------
-------------------------

haveagood committed:
[BE] feat : 즐겨찾기 기능 구현

- bookmarks table에 존재하는 user의 name으로 save요청이 들어오면
  properties_id와 user_id를 저장
- 삭제 요청은 bookmarks에서 삭제하도록 구현
- 기존의 properties에서 즐겨찾기 되어있는 숙소는 saved로 표시될 수
  있도록 구현

-------------------------
-------------------------

haveagood committed:
[BE] feat : saved 즐겨찾기 api 구현 완료

- user에 따른 별도의 saved 리스트를 가지고 있도록 구현
- 다른 userId로 요청을 하면 해당 user고유의 savedList가 뜨도록 구현
- 특정 user의 savedList를 볼 수 있는 api 구현
- saved된 properties를 unsaved하는 기능 구현

-------------------------
-------------------------

hanurii committed:
[BE] feat : 예약하기 기능 구현

hanurii committed:
[BE] feat : 예약취소, 예약목록 기능 구현

hanurii committed:
[BE] feat : 상세 페이지 기능 구현

hanurii committed:
[BE] feat : 예약 시 calendar 테이블에 데이터 추가 구현

- issue : #60

hanurii committed:
[BE] feat : 예약 취소 시 해당하는 calendar 데이터 값도 삭제

- issue : #60

hanurii committed:
[BE] feat : 지도 API 구현 완료

- issue : #8

hanurii committed:
[BE] feat : saved 즐겨찾기 추가 기능 구현

- issue : #9

hanurii committed:
[BE] feat : 즐겨찾기 시 properties 의 saved 변수 true로 변경, saved 전체 목록 출력 구현

- issue : #9

hanurii committed:
[BE] feat : 즐겨찾기 삭제 기능 구현

- issue : #9

hanurii committed:
[BE] feat : 예약 시 reservable = false 로 변경하는 기능 구현

- issue : #60

hanurii committed:
[BE] modify : 예약 취소 시 propertyId 에 해당하는 숙소가 사라지는 이슈 해결

- issue : #60

hanurii committed:
[BE] feat : 즐겨찾기 삭제 시 saved = false로 변경하는 기능 구현

- issue : #9

hanurii committed:
Merge branch 'feat/be/api-merge' into feat/be/api-develop-hamill

# Conflicts:
#	be/out/production/resources/application.properties
#	be/out/production/resources/data.sql
#	be/src/main/java/com/airbnb3/codesquad/airbnb3/controller/AirbnbControllerHamill.java
#	be/src/main/java/com/airbnb3/codesquad/airbnb3/dao/PropertiesDaoHamill.java
#	be/src/main/java/com/airbnb3/codesquad/airbnb3/service/AirbnbServiceHamill.java
#	be/src/main/resources/application.properties
#	be/src/main/resources/data.sql

hanurii committed:
[BE] modify : 코드 병합 후 hamill API 돌아가게 수정

hanurii committed:
[BE] modify : url parameter 변경, 예약 시 1달 이상 숙박했을 때 calendar 에 제대로 값이 들어가지 않는 이슈 해결, schema에 가격 관련 변수들 타입을 decimal로 변경

hanurii committed:
[BE] feat : 숙소 예약 시 숙소에 관한 예약 상세페이지 출력

- issue : #60

hanurii committed:
[BE] modify : schmea에 있는 가격 관련 컬럼 타입 변경(double -> decimal)

hanurii committed:
[BE] feat  : 예약하기, 예약취소, 예약 목록 출력 기능 재구현

- 예약 시 예약한 숙소의 예약 정보를 보여준다

- issue : #60

hanurii committed:
[BE] feat : ReservationsDto 만들어서 예약했을 때 보내주는 JSON 응답 객체 구현

- issue : #60

hanurii committed:
Merge branch 'feat/be/api-merge' into feat/be/deploy-hamill

# Conflicts:
#	be/src/main/resources/application.properties

haveagood committed:
[BE] refactor : api method 변경

- test를 위해 getmapping으로 받았던 메소드들을 역할에 맞게 수정

-------------------------
-------------------------

hanurii committed:
[BE] feat : 예약 목록이 응답하는 JSON 객체 변경

- issue : #60

hanurii committed:
[BE] modify : saved/id 를 saved/propertyId로 변경, 즐겨찾기 추가,삭제 시 메세지 객체를 응답

- issue : #9

hanurii committed:
Merge branch 'feat/be/api-merge' into feat/be/deploy-hamill

hanurii committed:
[BE] modify : 즐겨찾기 추가,삭제와 예약 취소 시 전송되는 메세지 수정

- status code와 message가 뒤바뀐채로 전송되고 있어서 수정함

hanurii committed:
[BE] feat : 중복 예약 시 400 에러 메세지 반환

- issue : #60

haveagood committed:
[BE] feat : search 기능 추가

- title 부분에 대한 검색을 sql에서 검색하여 결과값을 출력하도록 구현
- 필터 기능의 하나로 추가하였음

-------------------------
-------------------------

hanurii committed:
[BE] feat : JWT와 인터셉터 구현

- GitHub 인증 허가 시 리소스를 서버로 받아오고 받은 리소스를 JWT로 변환하여 클라이언트에게 전달한다.
- 클라이언트는 로그인이 필요한 서비스 이용 시 전달받은 JWT를 헤더에 담아서 서버에게 보낸다.
- 서버는 JWT의 유효 검사를 진행 후(prehandle) 적합하면 API를 응답해준다.

- issue : #111

hanurii committed:
Merge pull request #119 from codesquad-member-2020/feat/be/deploy-hamill

배포 작업물을 머지합니다

hanurii committed:
[BE] feat : GitHub Action 구현 완료

hanurii committed:
[BE] fix : statusCode와 message가 뒤바뀐 응답이여서 수정

hanurii committed:
Merge pull request #121 from codesquad-member-2020/feat/be/hotfix

[BE] fix : statusCode와 message가 뒤바뀐 응답이여서 수정

hanurii committed:
Merge pull request #122 from codesquad-member-2020/feat/be/api-develop-hamill

[BE] JWT 인터셉터 기능 PR MERGE

hanurii committed:
[BE] fix : 잘못된 도메인 주소 변경

hanurii committed:
[BE] feat : 인터셉터로 걸러지지 않게 처리

hanurii committed:
Merge pull request #123 from codesquad-member-2020/feat/be/hotfix

불필요한 인터셉터 기능을 죽여놓음

hanurii committed:
Merge pull request #126 from codesquad-member-2020/feat/be/api-merge

[BE] feat : search 기능 추가
